include/knowhere/sparse_utils.h

```diff
@@ -0,0 +1,226 @@
+// Copyright (C) 2019-2023 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software valributed under the License
+// is valributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <type_traits>
+#include <vector>
+
+#include "knowhere/operands.h"
+
+namespace knowhere::sparse {
+
+// integer type in SparseRow
+using table_t = uint32_t;
+// type used to represent the id of a vector in the index interface.
+// this is same as other index types.
+using label_t = int64_t;
+
+template <typename T>
+struct IdVal {
+    table_t id;
+    T val;
+
+    IdVal() = default;
+    IdVal(table_t id, T val) : id(id), val(val) {
+    }
+
+    inline friend bool
+    operator<(const IdVal& lhs, const IdVal& rhs) {
+        return lhs.val < rhs.val || (lhs.val == rhs.val && lhs.id < rhs.id);
+    }
+    inline friend bool
+    operator>(const IdVal& lhs, const IdVal& rhs) {
+        return !(lhs < rhs);
+    }
+
+    inline friend bool
+    operator==(const IdVal& lhs, const IdVal& rhs) {
+        return lhs.id == rhs.id && lhs.val == rhs.val;
+    }
+};
+
+template <typename T>
+class SparseRow {
+    static_assert(std::is_same_v<T, fp32>, "SparseRow supports float only");
+
+ public:
+    SparseRow() = default;
+
+    SparseRow(const SparseRow& other) {
+        copy_from(other);
+    }
+
+    SparseRow&
+    operator=(const SparseRow& other) {
+        copy_from(other);
+        return *this;
+    }
+
+    ~SparseRow() {
+        if (own_data && data != nullptr) {
+            delete[] data;
+        }
+    }
+
+    size_t
```

Comment:
move constructor and move assignment operator please :)

Reply:
Done